lighthouse-core/gather/driver.js

```diff
@@ -630,6 +630,60 @@ class Driver {
     return new LHElement(targetNode, this);
   }
 
+  /**
+   * @param {string} handle
+   * @param {{timeout: number}=} options,
+   * @return {Promise<string>}
+   */
+  async readIOStream(handle, options = {timeout: 5000}) {
+    const startTime = Date.now();
+
+    let ioResponse;
+    let data = '';
+    while (!ioResponse || !ioResponse.eof) {
+      if (Date.now() - startTime > options.timeout) {
+        throw new Error('Waiting for the end of the IO stream exceeded the allotted time.');
+      }
+      ioResponse = await this.sendCommand('IO.read', {handle});
+      data = data.concat(ioResponse.data);
+    }
+
+    if (ioResponse.base64Encoded) {
+      data = Buffer.from(data, 'base64').toString('utf-8');
```

Comment:
this seems like it would break if there is more than one chunk of base64 data. each chunk get's deserialized and saved to the same data variable.

I think if one io response is encoded, all will be. could just store the data string as it comes over the protocol, and then at the end of the function check the encoded flag and do a `Buffer.from`

EDIT: actually, base64 isn't a streamable format (can't concat the encoding data and decode it..), so you gotta decode each chunk as you go. so just do Buffer.from on each `ioResponse` before adding to `data`.

Reply:
So the data isn't encoded in base64 before splitting into chunks. Each chunk is encoded in base64 individually after splitting.

Am I understanding this correctly?

Reply:
I have no idea, what does the chromium source suggest?

Whatever you find out it'd be good to document that in the protocol definition file

Reply:
https://source.chromium.org/chromium/chromium/src/+/master:content/browser/devtools/devtools_stream_file.cc;l=86-127;drc=92a46414765f4873251cf9ecefbef130648e2af8;bpv=1;bpt=1

Looks like the chunks are encoded after splitting.